### DISCUSSION TITLE
A 'required' class

### DISCUSSION THREAD

**Discussion by royberris** [dotnet/csharplang#6814] [Language Ideas]
When C#11 came around the only thing I was really excited about was the required keyword. Nullability was a great improvement but was a pain sometimes before the required keyword. After using it for some weeks I've noticed that, while it works fine, it is a bit clumsy.

```csharp
public class Foo
{
    public required string Bar { get; set; }
    public string? Baz { get; set; }
}
```

Looking at above example. `Bar` should always have a value on initialization. `Baz` doesn't, because it isn't required. While required is related to nullability they aren't equal. For some reason I can define a required nullable member. Which can still be null but just needs to be acted on, it is confusing. The clumsy part is when you get bigger classes that have 10, 15+ properties. It is very hard to read through the list of properties to see what is required, what is null, etc.

I am a big fan of thinking about the code you write. Making sure things are null for a reason, not just because one is a lazy programmer. But I am also a fan of less syntax.

A `required class` would bring the two together where a requirement is implicitly the same as its nullability.

```csharp
public required class Foo
{
    public string Bar { get; set; }
    public string? Baz { get; set; }
}
```

In above example `Bar` needs to be initialized, because it is not nullable. Simple as that. 

----------------

I understand it couldn't be implemented like this (without the `required` keyword) because of backwards compatibility. An alternative would be a flag in the `csproj` just as we have with `<Nullable>enable</Nullable>`. Making every non-nullable property automatically required.

_Note: I have no idea on if this is even possible to implement, just sharing thoughts and let the experts decide_

**Comment by CyrusNajmabadi** [comment]
> For some reason I can define a required nullable member.

 Required and nullable are two orthogonal concepts. Nullable simply states of the value can be null or not. Required states if the value must be provided when constructed or not.  It's very reasonable, and normal, to have required nullable members. That simply says they must be provided, but are allowed to be given the value null. 

Required, in this way, is similar to an unordered parameter list for the type.  Just like a parameter list, you just provide all the parameters. The exception are optional parameters. But those correspond to non -required members.

**Reply by TahirAhmadov** [reply]
This is definitely a good candidate for the BIG RED BOLD FAQ section on top of the discussions page :)

**Comment by jnm2** [comment]
> Which can still be null but just needs to be acted on, it is confusing.

This is a really clear way of stating it. Since there's no contradiction in a class author wanting maybe-null things to be acted on, and you seem to have a good grasp of the concept, what makes it confusing?

**Comment by DavidArno** [comment]
> For some reason I can define a required nullable member. Which can still be null but just needs to be acted on, it is confusing.

Consider a variation on your example:
```cs
public class X
{
    public required int Y { get; set; }
}
```
I have defined `Y` as required. It can still be `0`, but it needs to be explicitly set to that. Is that also confusing? 

Whilst I understand your argument that enforcing `required` on a property that can accept the default value as a valid value is undesirable, I really dislike the idea that `null` be a special case default value that gets excluded from being `required`. In your example:
```cs
public required class Foo
{
    public string Bar { get; set; }
    public string? Baz { get; set; }
}
```
I like the idea of being able to specify `required` at the class level. But the principle of least astonishment (POLA) comes into play here for me. The POLA behaviour has to be that all the properties are required. Excluding `Baz` simply because `null` is a valid initialization value feels like a clear POLA violation.

**Reply by jnm2** [reply]
I can't stop hearing POLA as "principle of least action."

**Comment by KennethHoff** [comment]
I would love to see this introduced as well, but with its introduction I think we should probably have a way to _exclude_ certain properties, say `optional`. Although, it wouldn't make sense to have `optional` on its own outside of a required class. 

```csharp
public required class Foo
{
    public string Bar { get; set; }
    public optional string? Baz { get; set; }
}
```

Additionally, would all nested classes also be required by default, or is it just syntactic sugar for applying required to all properties on the "active" class?


```csharp
public class ChildFoo : Foo
{
    public string ChildBar {get; set; }
}
```

`ChildBar` in this case would not be required.

(Additionally, I imagine this would all also obviously extend to structs and records(and record structs) as well - not just classes)
